pkg/driver/camera/camera_windows.cpp

```
@@ -422,30 +437,75 @@ HRESULT SampleGrabberCallback::BufferCB(double sampleTime, BYTE* buf, LONG len)
+    const BYTE* uv = buf + nPix;
+    BYTE* u = gobuf + nPix;
+    BYTE* v = u + nPix / 4;
+    
+    int uvSize = nPix / 4;
+    int i = 0;
+    
+    // Process 4 pixels at a time
+    for (; i < uvSize - 3; i += 4)
+    {
+      u[i] = uv[i * 2];
+      v[i] = uv[i * 2 + 1];
+      u[i + 1] = uv[(i + 1) * 2];
+      v[i + 1] = uv[(i + 1) * 2 + 1];
+      u[i + 2] = uv[(i + 2) * 2];
+      v[i + 2] = uv[(i + 2) * 2 + 1];
+      u[i + 3] = uv[(i + 3) * 2];
+      v[i + 3] = uv[(i + 3) * 2 + 1];
+    }
+    
+    // Handle remaining pixels
+    for (; i < uvSize; i++)
+    {
+      u[i] = uv[i * 2];
+      v[i] = uv[i * 2 + 1];
+    }
   }
-  int yi = 0;
-  int cbi = cam_->width * cam_->height;
-  int cri = cbi + cbi / 2;
-  // Pack as I422
-  for (int y = 0; y < cam_->height; ++y)
+  else if (abs(len - expectedYUY2) <= 10)
   {
-    int j = y * cam_->width * 2;
-    for (int x = 0; x < cam_->width / 2; ++x)
+    // YUY2 → I420: Extract Y, subsample and average UV vertically
+    BYTE* y = gobuf;
+    BYTE* u = gobuf + nPix;
+    BYTE* v = u + nPix / 4;
+    
+    for (int row = 0; row < cam_->height; row += 2)
     {
-      gobuf[yi] = buf[j];
-      gobuf[cbi] = buf[j + 1];
-      gobuf[yi + 1] = buf[j + 2];
-      gobuf[cri] = buf[j + 3];
-      j += 4;
-      yi += 2;
-      cbi++;
-      cri++;
+      const BYTE* src1 = buf + row * cam_->width * 2;
+      const BYTE* src2 = src1 + cam_->width * 2;
+      
+      for (int col = 0; col < cam_->width; col += 2)
+      {
+        y[0] = src1[0];
+        y[1] = src1[2];
+        y[cam_->width] = src2[0];
+        y[cam_->width + 1] = src2[2];
+        y += 2;
+        
+        *u++ = (src1[1] + src2[1]) / 2;
+        *v++ = (src1[3] + src2[3]) / 2;
+        
+        src1 += 4;
+        src2 += 4;
+      }
+      y += cam_->width;
     }
   }
+  else
+  {
+    fprintf(stderr, "Unexpected buffer size: %d (expected NV12=%d or YUY2=%d)\n", 
+            len, expectedNV12, expectedYUY2);
+    return S_OK;
```

I think the whole program should fail instead of ignoring the data if an incompatible format is selected.

---

I agree